### ISSUE TITLE
element-web not usable on 3GB RAM ARM64 device due to excessive memory usage

### ISSUE THREAD

**Issue by ell1e** [vector-im/element-web#23788]
### Steps to reproduce

1. Use 3GB RAM ARM64 device like the PinePhone (Allwinner/3GB variant) in Firefox
2. Make sure earlyoom is active
3. element-web tab keeps crashing seconds after loading because it uses too much memory

Since telegram-desktop works perfectly fine (also runs infinitely smoother btw) and Discord web at least crashes more rarely even though I'm in probably 2x-5x the chat rooms on Discord to Element, this doesn't look reasonable in comparison to competitors.

#### What did you expect?
Works
#### What happened instead?
Constantly getting its tab killed due to excessive memory use

### Operating system

postmarketOS (based on Alpine Linux)

### Browser information

Firefox ESR 91.13.0esr

### URL for webapp

app.element.io

### Application version

crashes too early to reasonably check

### Homeserver

matrix.org

### Will you send logs?

No

**Comment by robintown** [comment]
This will potentially be solved by Sliding Sync.

**Comment by t3chguy** [comment]
Non-desktop environment is experimental as per support policy

**Comment by ell1e** [comment]
I just terminated element-desktop after using short of 700MB. Unless the code base difference to element-web is the reason for the huge discrepancy, I'd suggest this is also affecting desktops since that's a pretty sizeable amount to gobble up on a regular lower spec computer.

**Comment by DeezyTheFox** [comment]
> Edit: after restarting it to see how much of this was leaks over time, it actually used 900MB right away. Wow. Eventually it went down to below 500MB, but that peak is pretty extreme for a chat app.

I'll just put it here
![2022-11-24 16 30 33](https://user-images.githubusercontent.com/19533171/207855523-e7065455-dc1e-4a3c-a271-ae5ab41d381a.png)

**Comment by mpeter50** [comment]
On desktop I only use Element-Web in browser (Firefox), and according to the about:processes page it easily consumes 1-2 GB of RAM, sometimes 3. Not right after loading the client, but a few hours or a day later.
This is particularly bad when I have several tabs of the Element client open, and I do that often so that I have different rooms open for different chats where I dont want to forget that I'm waiting for a response, but I don't want to turn on receiving notifications either.

As an example, this Element client was opened approximately 6 hours ago:
![image](https://user-images.githubusercontent.com/83356418/210432355-e1dfd50a-3dcb-4f61-a6a6-b6ee4839f380.png)

The screenshot is from the about:processes built-in page of Firefox.

Devs: if I can help in any way, even with performance profiles or other involved things, I'm happy to help!

**Comment by JeanPaulLucien** [comment]
> > Edit: after restarting it to see how much of this was leaks over time, it actually used 900MB right away. Wow. Eventually it went down to below 500MB, but that peak is pretty extreme for a chat app.
> 
> I'll just put it here ![2022-11-24 16 30 33](https://user-images.githubusercontent.com/19533171/207855523-e7065455-dc1e-4a3c-a271-ae5ab41d381a.png)

Oh. It was here. But they did not trust me.

**Comment by JeanPaulLucien** [comment]
> Devs: if I can help in any way, even with performance profiles or other involved things, I'm happy to help!

Yeah, check my issues. But you will not make a Profile if you have 8 GB ram or less.

**Comment by paboum** [comment]
Instead of trying to retroactively catch memory leaks, it might be more prudent to guarantee that reasonable practices are followed in the code.

https://www.electronjs.org/docs/latest/tutorial/performance presents a recommendations checklist to follow. Perhaps one of the devs could prepare a report on how well are they followed, or even write down the action items? Then, the users won't need to facilitate bug tracking and could observe the progress instead.

**Comment by ell1e** [comment]
On my laptop which has 16GB RAM, the element web tab now regularly gets nuked by Firefox (unloaded) due to the memory usage. The only difference to the small and weaker device is that it's not an uncontrolled crash, but it's a problem nevertheless, even if just to keep notifications working and to remain reachable.

**Comment by ell1e** [comment]
Sadly, with recent updates this seems to have gotten *even worse* 😢 with a 3GB constraint often it will out of memory crash on initial load before I can even send a single message, while it used to at least work for some minutes. And even on my normal specs topic the tab crashes every hour or so. It's quite sad

**Comment by mpeter50** [comment]
It is indded sad, but I think there are more appropriate words for it.

A webapp should not use more than 1 GB of RAM, and even that is too many.

But to also be helpful: I have a(n untested) suspicion that high memory usage is related to being joined to a lot of rooms (a few dozen). You may try leaving most of the public ones that you dont regularly use, and see if it improves.
Go through the People and the Rooms setion on the sidebar of the "Home" virtual space, but also any other space you have because rooms from those are not shown in the former list.

**Comment by ell1e** [comment]
I appreciate your remark, although my commentary of it is more aimed at the devs: just in case anyone would ever more officially suggest that as a serious workaround, IMHO joining less rooms would not be an acceptable solution even if it happened to help. I'm in about 10x as many rooms in discord and it runs fine on the same device (well, it lags horribly, but it doesn't OOM crash after barely a minute). Nevertheless, it seems to maybe be an interesting pointer as to where the problem might be.

**Comment by paboum** [comment]
> With it your client doesn't need to keep every single joined room loaded.

@t3chguy: I fail to understand your logic here. If, even without sliding-sync, I can close the tab and then reopen it later, so the client collects all the data it needs from the local storage and the homeserver, how come it _needs_ to store all that in RAM before I close the tab?

Perhaps some kind of garbage collector would help, eh? Let's say the user decides to only allocate 1 GB of memory for client cache. The collector would then periodically check if the limit is not exceeded and free up decrypted messages, downloaded media and other stuff that sticks out and then free it, in the longest-time-not-seen order. Just look how tabs are implemented in modern browsers - I can easily have 100 tabs open and keep my firefox somewhere under 1 GB of actual RAM usage (I just don't display all of them every time I open the browser).

Or, even better, allow users to decide - e.g. this is a large room which I read seldomly, let's unload it from my memory and only load it when I explicitly want it. Still need a way of notifications though so it's up to you how much of this could be done. This, keeping the browser tabs analogy, could be similar to browser task manager - displaying memory consumption of each tab and allowing to "unload tab and kill process". I guess a similar thing could be tricky since javascript is a single threaded language, but surely you can take the idea from here.
